server/controllers/finance/reports/unbalanced_invoice_payments/index.js

```diff
@@ -0,0 +1,133 @@
+const _ = require('lodash');
+const ReportManager = require('../../../../lib/ReportManager');
+const db = require('../../../../lib/db');
+const util = require('../../../../lib/util');
+// path to the template to render
+const TEMPLATE = './server/controllers/finance/reports/unbalanced_invoice_payments/report.handlebars';
+
+exports.document = build;
+
+/**
+ * Actually builds the open debtor report.
+ */
+function build(req, res, next) {
+  const qs = _.extend(req.query, {
+    footerRight : '[page] / [toPage]',
+    footerFontSize : '7',
+  });
+
+  const metadata = _.clone(req.session);
+
+  let report;
+  try {
+    report = new ReportManager(TEMPLATE, metadata, qs);
+  } catch (e) {
+    return next(e);
+  }
+  const { dateFrom, dateTo } = req.query;
+  return getBalances(req.query).then(result => {
+    return report.render(
+      _.extend({}, result, { dateFrom, dateTo })
+    );
+  })
+    .then((compiledReport) => {
+      res.set(compiledReport.headers).send(compiledReport.report);
+    })
+    .catch(next);
+}
+
+// invoice payements balance
+async function getBalances(options) {
+
+  const { dateFrom, dateTo } = options;
+  const debtorsSql = `
```

Comment:
Can't you put this SQL in the stored procedure?  That way, you won't need to loop through a large JS array?